File: docs/assets/notable-windows.js
(function initialiseNotableWindows(root, factory) {
  const api = factory();
  root.AgamCsNotableWindows = api;
  if (typeof module === 'object' && module.exports) module.exports = api;
}(globalThis, () => {
  const ANALYSIS_VERSION = 'agamcs-notable-windows-v1';
  const SCHEMA_VERSION = 1;
  const COORDINATE_CONVENTION = '1-based inclusive';
  const WINDOW_SIZE = 100;
  const TOP_WINDOWS = 5;
  const SNP_ACCESSIBILITY_THRESHOLD = 0.8;
  const TIE_TOLERANCE = 1e-12;

  function clip(start, end, lower, upper) {
    const clippedStart = Math.max(start, lower);
    const clippedEnd = Math.min(end, upper);
    return clippedStart <= clippedEnd ? [clippedStart, clippedEnd] : null;
  }

  function merge(intervals) {
    const merged = [];
    [...intervals].sort((left, right) => left[0] - right[0] || left[1] - right[1]).forEach(([start, end]) => {
      if (!merged.length || start > merged[merged.length - 1][1] + 1) merged.push([start, end]);
      else merged[merged.length - 1][1] = Math.max(merged[merged.length - 1][1], end);
    });
    return merged;
  }

  function subtract(intervals, removed) {
    const output = [];
    merge(intervals).forEach(([start, end]) => {
      let cursor = start;
      for (const [removedStart, removedEnd] of merge(removed)) {
        if (removedEnd < cursor) continue;
        if (removedStart > end) break;
        if (removedStart > cursor) output.push([cursor, Math.min(end, removedStart - 1)]);
        cursor = Math.max(cursor, removedEnd + 1);
        if (cursor > end) break;
      }
      if (cursor <= end) output.push([cursor, end]);
    });
    return output;
  }

  function finiteMean(values) {
    const finite = [...values].map((value) => value == null ? Number.NaN : Number(value)).filter(Number.isFinite);
    return { mean: finite.length ? finite.reduce((sum, value) => sum + value, 0) / finite.length : null, count: finite.length };
  }

  function normaliseResult(source) {
    const chromosome = String(source?.chromosome || '');
    const { start, end } = source || {};
    const values = source?.values || {};
    const required = ['Cs', 'snp_density', 'status'];
    if (!chromosome || !Number.isSafeInteger(start) || !Number.isSafeInteger(end) || start < 1 || end < start) {
      throw new Error('Notable-window analysis requires valid inclusive query coordinates.');
    }
    if (required.some((name) => values[name] == null)) throw new Error('Notable-window analysis requires Cs, snp_density, and status arrays.');
    const expected = end - start + 1;
    if (required.some((name) => values[name].length !== expected)) throw new Error('Notable-window arrays must match the inclusive query length.');
    return { chromosome, start, end, values };
  }

  function featureSegments(annotation, result) {
    if (!annotation?.transcript_id) return { transcript: null, segments: [] };
    if (String(annotation.chromosome) !== result.chromosome) throw new Error('Selected transcript and query must use the same chromosome.');
    const { start, end } = annotation;
    const strand = Number(annotation.strand ?? 1);
    if (!Number.isSafeInteger(start) || !Number.isSafeInteger(end) || start < 1 || end < start || ![-1, 1].includes(strand)) throw new Error('Selected transcript requires valid coordinates and strand.');
    const transcriptInterval = clip(start, end, result.start, result.end);
    const exons = merge((annotation.exons || []).map((exon) => clip(Number(exon.start), Number(exon.end), result.start, result.end)).filter(Boolean));
    let { cds_start: cdsStart, cds_end: cdsEnd } = annotation;
    if ((cdsStart == null) !== (cdsEnd == null)) throw new Error('Selected transcript CDS bounds must both be present or absent.');
    let coding = [];
    if (cdsStart != null) {
      cdsStart = Number(cdsStart); cdsEnd = Number(cdsEnd);
      if (!Number.isSafeInteger(cdsStart) || !Number.isSafeInteger(cdsEnd) || cdsStart > cdsEnd) throw new Error('Selected transcript CDS start cannot exceed CDS end.');
      coding = merge(exons.map(([exonStart, exonEnd]) => clip(exonStart, exonEnd, cdsStart, cdsEnd)).filter(Boolean));
    }
    const segments = [];
    [...exons].sort((left, right) => strand === -1 ? right[0] - left[0] || right[1] - left[1] : left[0] - right[0] || left[1] - right[1]).forEach(([exonStart, exonEnd], index) => {
      const number = index + 1;
      const exonCoding = merge(coding.map(([codingStart, codingEnd]) => clip(exonStart, exonEnd, codingStart, codingEnd)).filter(Boolean));
      if (!exonCoding.length) { segments.push([exonStart, exonEnd, `Non-coding exon ${number}`]); return; }
      subtract([[exonStart, exonEnd]], exonCoding).forEach(([utrStart, utrEnd]) => {
        const fivePrime = strand === 1 ? utrEnd < cdsStart : utrStart > cdsEnd;
        segments.push([utrStart, utrEnd, `${fivePrime ? '5′ UTR' : '3′ UTR'} (exon ${number})`]);
      });
      exonCoding.forEach(([codingStart, codingEnd]) => segments.push([codingStart, codingEnd, `CDS (exon ${number})`]));
    });
    if (transcriptInterval) subtract([transcriptInterval], exons).forEach(([intronStart, intronEnd]) => segments.push([intronStart, intronEnd, 'Intron']));
    return { transcript: { gene_id: annotation.id ?? null, transcript_id: String(annotation.transcript_id), strand }, segments: segments.sort((left, right) => left[0] - right[0] || left[1] - right[1] || left[2].localeCompare(right[2])) };
  }

  function windowFeature(start, end, transcript, segments) {
    if (!transcript) return 'No selected transcript';
    const labels = [];
    segments.forEach(([segmentStart, segmentEnd, label]) => {
      if (segmentStart <= end && segmentEnd >= start && !labels.includes(label)) labels.push(label);
    });
    return labels.length ? labels.join('; ') : 'Outside selected transcript';
  }

  function windowRecord(result, start, end, transcript, segments) {
    const offset = start - result.start;
    const length = end - start + 1;
    const cs = result.values.Cs.slice(offset, offset + length);
    const snp = result.values.snp_density.slice(offset, offset + length);
    const status = result.values.status.slice(offset, offset + length);
    const accessible = [...status].map((value) => (Number(value) & 1) === 1);
    const accessibleBases = accessible.filter(Boolean).length;
    const csMean = finiteMean(cs);
    const snpMean = finiteMean([...snp].filter((_value, index) => accessible[index]));
    const accessibleFraction = accessibleBases / length;
    return {
      chromosome: result.chromosome, start, end, total_bases: length,
      finite_cs_bases: csMean.count, mean_cs: csMean.mean,
      accessible_bases: accessibleBases, accessible_fraction: accessibleFraction,
      finite_accessible_snp_bases: snpMean.count, mean_accessible_snp_density: snpMean.mean,
      snp_density_eligible: accessibleFraction >= SNP_ACCESSIBILITY_THRESHOLD && snpMean.mean != null,
      selected_transcript_feature: windowFeature(start, end, transcript, segments),
    };
  }

  function compareWindows(left, right, metric, descending = false) {
    const difference = left[metric] - right[metric];
    if (Math.abs(difference) > TIE_TOLERANCE) return descending ? -difference : difference;
    return left.start - right.start;
  }

  function analyzeNotableWindows(sourceResult, annotation = null, { windowSize = WINDOW_SIZE, topWindows = TOP_WINDOWS } = {}) {
    if (!Number.isSafeInteger(windowSize) || windowSize < 1) throw new Error('Window size must be a positive integer.');
    if (!Number.isSafeInteger(topWindows) || topWindows < 1) throw new Error('Top-window count must be a positive integer.');
    const result = normaliseResult(sourceResult);
    const { transcript, segments } = featureSegments(annotation, result);
    const windows = [];
    for (let start = result.start; start <= result.end; start += windowSize) windows.push(windowRecord(result, start, Math.min(start + windowSize - 1, result.end), transcript, segments));
    const highestMeanCs = windows.filter((window) => window.mean_cs != null).sort((left, right) => compareWindows(left, right, 'mean_cs', true)).slice(0, topWindows);
    const lowestMeanSnpDensity = windows.filter((window) => window.snp_density_eligible).sort((left, right) => compareWindows(left, right, 'mean_accessible_snp_density')).slice(0, topWindows);
    return {
      schema_version: SCHEMA_VERSION, analysis_version: ANALYSIS_VERSION,
      coordinate_convention: COORDINATE_CONVENTION, window_size: windowSize, top_windows: topWindows,
      snp_accessibility_threshold: SNP_ACCESSIBILITY_THRESHOLD,
      query: { chromosome: result.chromosome, start: result.start, end: result.end },
      selected_transcript: transcript, windows,
      highest_mean_cs_windows: highestMeanCs, lowest_mean_snp_density_windows: lowestMeanSnpDensity,
    };
  }

  return { ANALYSIS_VERSION, COORDINATE_CONVENTION, SCHEMA_VERSION, SNP_ACCESSIBILITY_THRESHOLD, TOP_WINDOWS, WINDOW_SIZE, analyzeNotableWindows };
}));

File: AgamCs/notable_windows.py
"""Deterministic, QC-aware notable-window analysis for exact query data.

The matching browser implementation lives in ``docs/assets/notable-windows.js``.
Both are tested against ``tests/fixtures/notable-windows-v1-cases.json``.
"""

from __future__ import annotations

import math
from collections.abc import Mapping
from functools import cmp_to_key


ANALYSIS_VERSION = 'agamcs-notable-windows-v1'
SCHEMA_VERSION = 1
COORDINATE_CONVENTION = '1-based inclusive'
WINDOW_SIZE = 100
TOP_WINDOWS = 5
SNP_ACCESSIBILITY_THRESHOLD = 0.8
TIE_TOLERANCE = 1e-12


def _clip(start: int, end: int, lower: int, upper: int):
    start, end = max(start, lower), min(end, upper)
    return (start, end) if start <= end else None


def _merge(intervals):
    merged = []
    for start, end in sorted(intervals):
        if not merged or start > merged[-1][1] + 1:
            merged.append([start, end])
        else:
            merged[-1][1] = max(merged[-1][1], end)
    return [tuple(interval) for interval in merged]


def _subtract(intervals, removed):
    output = []
    for start, end in _merge(intervals):
        cursor = start
        for removed_start, removed_end in _merge(removed):
            if removed_end < cursor:
                continue
            if removed_start > end:
                break
            if removed_start > cursor:
                output.append((cursor, min(end, removed_start - 1)))
            cursor = max(cursor, removed_end + 1)
            if cursor > end:
                break
        if cursor <= end:
            output.append((cursor, end))
    return output


def _finite_mean(values):
    values = [float(value) for value in values if value is not None and math.isfinite(float(value))]
    return (sum(values) / len(values) if values else None), len(values)


def _normalise_result(source: Mapping) -> dict:
    chromosome = str(source.get('chromosome') or '')
    start, end = source.get('start'), source.get('end')
    values = source.get('values') or {}
    required = ('Cs', 'snp_density', 'status')
    if not chromosome or not isinstance(start, int) or not isinstance(end, int) or start < 1 or end < start:
        raise ValueError('Notable-window analysis requires valid inclusive query coordinates.')
    if any(name not in values for name in required):
        raise ValueError('Notable-window analysis requires Cs, snp_density, and status arrays.')
    expected = end - start + 1
    if any(len(values[name]) != expected for name in required):
        raise ValueError('Notable-window arrays must match the inclusive query length.')
    return {'chromosome': chromosome, 'start': start, 'end': end, 'values': values}


def _feature_segments(annotation: Mapping | None, result: Mapping):
    """Return non-overlapping selected-transcript labels clipped to the query."""
    if not annotation or not annotation.get('transcript_id'):
        return None, []
    if str(annotation.get('chromosome')) != result['chromosome']:
        raise ValueError('Selected transcript and query must use the same chromosome.')
    start, end = annotation.get('start'), annotation.get('end')
    strand = int(annotation.get('strand', 1))
    if not isinstance(start, int) or not isinstance(end, int) or start < 1 or end < start or strand not in (-1, 1):
        raise ValueError('Selected transcript requires valid coordinates and strand.')
    transcript = _clip(start, end, result['start'], result['end'])
    exons = _merge(
        clipped for exon in annotation.get('exons') or ()
        if (clipped := _clip(int(exon['start']), int(exon['end']), result['start'], result['end']))
    )
    cds_start, cds_end = annotation.get('cds_start'), annotation.get('cds_end')
    if (cds_start is None) != (cds_end is None):
        raise ValueError('Selected transcript CDS bounds must both be present or absent.')
    coding = []
    if cds_start is not None:
        cds_start, cds_end = int(cds_start), int(cds_end)
        if cds_start > cds_end:
            raise ValueError('Selected transcript CDS start cannot exceed CDS end.')
        coding = _merge(
            clipped for exon_start, exon_end in exons
            if (clipped := _clip(exon_start, exon_end, cds_start, cds_end))
        )
    ordered_exons = sorted(exons, reverse=strand == -1)
    segments = []
    for number, (exon_start, exon_end) in enumerate(ordered_exons, start=1):
        exon_coding = _merge(
            clipped for coding_start, coding_end in coding
            if (clipped := _clip(exon_start, exon_end, coding_start, coding_end))
        )
        if not exon_coding:
            segments.append((exon_start, exon_end, f'Non-coding exon {number}'))
            continue
        for utr_start, utr_end in _subtract([(exon_start, exon_end)], exon_coding):
            is_five_prime = (utr_end < cds_start) if strand == 1 else (utr_start > cds_end)
            label = '5′ UTR' if is_five_prime else '3′ UTR'
            segments.append((utr_start, utr_end, f'{label} (exon {number})'))
        for coding_start, coding_end in exon_coding:
            segments.append((coding_start, coding_end, f'CDS (exon {number})'))
    if transcript:
        segments.extend((start, end, 'Intron') for start, end in _subtract([transcript], exons))
    return {
        'gene_id': annotation.get('id'),
        'transcript_id': str(annotation['transcript_id']),
        'strand': strand,
    }, sorted(segments)


def _window_feature(window_start: int, window_end: int, transcript, segments) -> str:
    if transcript is None:
        return 'No selected transcript'
    labels = []
    for start, end, label in segments:
        if start <= window_end and end >= window_start and label not in labels:
            labels.append(label)
    return '; '.join(labels) if labels else 'Outside selected transcript'


def _window_record(result, start: int, end: int, transcript, segments) -> dict:
    offset = start - result['start']
    length = end - start + 1
    cs = result['values']['Cs'][offset:offset + length]
    snp = result['values']['snp_density'][offset:offset + length]
    status = result['values']['status'][offset:offset + length]
    accessible = [(int(value) & 1) == 1 for value in status]
    accessible_bases = sum(accessible)
    mean_cs, finite_cs_bases = _finite_mean(cs)
    mean_snp, finite_accessible_snp_bases = _finite_mean(
        value for value, is_accessible in zip(snp, accessible) if is_accessible
    )
    accessible_fraction = accessible_bases / length
    return {
        'chromosome': result['chromosome'],
        'start': start,
        'end': end,
        'total_bases': length,
        'finite_cs_bases': finite_cs_bases,
        'mean_cs': mean_cs,
        'accessible_bases': accessible_bases,
        'accessible_fraction': accessible_fraction,
        'finite_accessible_snp_bases': finite_accessible_snp_bases,
        'mean_accessible_snp_density': mean_snp,
        'snp_density_eligible': bool(
            accessible_fraction >= SNP_ACCESSIBILITY_THRESHOLD and mean_snp is not None
        ),
        'selected_transcript_feature': _window_feature(start, end, transcript, segments),
    }


def _compare_windows(left: Mapping, right: Mapping, metric: str, descending=False) -> int:
    """Compare means with a cross-runtime tolerance before coordinate tie-breaking."""
    left_value, right_value = left[metric], right[metric]
    if not math.isclose(left_value, right_value, rel_tol=0, abs_tol=TIE_TOLERANCE):
        if descending:
            return -1 if left_value > right_value else 1
        return -1 if left_value < right_value else 1
    return (left['start'] > right['start']) - (left['start'] < right['start'])


def analyze_notable_windows(source_result: Mapping, annotation: Mapping | None = None,
                            *, window_size: int = WINDOW_SIZE, top_windows: int = TOP_WINDOWS) -> dict:
    """Analyse exact query values in fixed windows anchored at the query start.

    ``window_size`` and ``top_windows`` are parameters for compact cross-language
    fixtures. Product callers use the fixed 100-base/five-window defaults.
    """
    if not isinstance(window_size, int) or window_size < 1:
        raise ValueError('Window size must be a positive integer.')
    if not isinstance(top_windows, int) or top_windows < 1:
        raise ValueError('Top-window count must be a positive integer.')
    result = _normalise_result(source_result)
    transcript, segments = _feature_segments(annotation, result)
    windows = []
    for start in range(result['start'], result['end'] + 1, window_size):
        windows.append(_window_record(
            result, start, min(start + window_size - 1, result['end']), transcript, segments,
        ))
    highest_cs = sorted(
        (window for window in windows if window['mean_cs'] is not None),
        key=cmp_to_key(lambda left, right: _compare_windows(left, right, 'mean_cs', descending=True)),
    )[:top_windows]
    lowest_snp = sorted(
        (window for window in windows if window['snp_density_eligible']),
        key=cmp_to_key(lambda left, right: _compare_windows(
            left, right, 'mean_accessible_snp_density',
        )),
    )[:top_windows]
    return {
        'schema_version': SCHEMA_VERSION,
        'analysis_version': ANALYSIS_VERSION,
        'coordinate_convention': COORDINATE_CONVENTION,
        'window_size': window_size,
        'top_windows': top_windows,
        'snp_accessibility_threshold': SNP_ACCESSIBILITY_THRESHOLD,
        'query': {key: result[key] for key in ('chromosome', 'start', 'end')},
        'selected_transcript': transcript,
        'windows': windows,
        'highest_mean_cs_windows': highest_cs,
        'lowest_mean_snp_density_windows': lowest_snp,
    }
